### PR TITLE
th v4

### PR DESCRIPTION
**Pull request by Logia556**
th v4

### CODE REVIEW

**Comment by Logia556** [comment]
please pull this pull request to add th_v4 watteco device

File: vendors/watteco/drivers/th_v4/package.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "TH",
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/package.json:2]: could we name the driver "th-v4" as now both v3 and v4 will be supported?

File: vendors/watteco/drivers/th_v4/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+const path = require("path");
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/webpack.config.js:1]: If you are obliged to have multiple js file for your decoder, and decided to use webpack. 

Here are the configurations to use: 

```
const path = require("path");

module.exports = {
  target: "node",
  mode: "production",
  entry: "./index.js",
  output: {
    filename: "main.js",
    path: path.resolve(__dirname, "."),
    library: "driver",
  },
};
```

in your case you could replace "index.js" by "th.js".

**Comment by mostafaibr** [comment]
All driver's files should be under the directory "/drivers/th_v4". so please move the files "decode.js", "batch.js", "standard.js".

File: vendors/watteco/models/th_v4/model.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Commercial name of the model
+name: TH
```
Comment (mostafaibr) [inline vendors/watteco/models/th_v4/model.yaml:2]: Could be named TH v4 to distinct the TH models. Now TH v3 and TH v4 would be supported.

File: vendors/watteco/drivers/th_v4/metadata.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "TH",
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/metadata.json:2]:
```suggestion
  "name": "th-v4",
```

File: vendors/watteco/drivers/th_v4/metadata.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  "version": "1.0.0",
+  "main": "index.js",
+  "description": "driver for TH sensor",
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/metadata.json:5]:
```suggestion
  "description": "driver for TH v4 sensor",
```

File: vendors/watteco/drivers/th_v4/package.json (hunk 1 above)
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/package.json:2]:
```suggestion
  "name": "th-v4",
```

File: vendors/watteco/drivers/th_v4/package.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  "version": "1.0.0",
+  "description": "TH captor npm packaged",
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/package.json:4]:
```suggestion
  "description": "TH v4 decoder npm packaged",
```

File: vendors/watteco/drivers/th_v4/package.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
+    "jest": "^29.6.1",
+    "webpack": "^5.88.2",
+    "webpack-cli": "^5.1.4"
+  },
+  "dependencies": {
+    "uglifyjs-webpack-plugin": "^2.2.0"
+  }
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/package.json:18]:
```suggestion
```

File: vendors/watteco/models/th_v4/model.yaml (hunk 3 above)
Comment (mostafaibr) [inline vendors/watteco/models/th_v4/model.yaml:2]:
```suggestion
name: TH v4
```

File: vendors/watteco/profiles/WATTECO-TH.1.0.2b_ETSI.yaml [hunk 8]
```diff
@@ -0,0 +1,94 @@
+# Please do not change the existing IDs (Some may follow the old format)
+# The name of the profile .yaml file should be the same as the following ID
+id: WATTECO-TH.1.0.2b_ETSI
```
Comment (mostafaibr) [inline vendors/watteco/profiles/WATTECO-TH.1.0.2b_ETSI.yaml:3]:
```suggestion
id: nke-watteco_RFGroup1_1.0.2b_classA
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+      "fPort": 125,
+      "recvTime": "2023-07-19T07:51:31.598957793Z"
+    },
+    "output": {
+      "data": {
+        "variable": "temperature",
+        "value": 29.53,
+        "date": "2023-07-19T07:51:31.598957793Z"
+      },
+      "warnings": [""]
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:26]:
```suggestion
      "warnings": []
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+      "recvTime": "2023-07-19T07:51:25.508306410Z"
+    },
+    "output": {
+      "data": {
+        "variable": "disposable_battery_voltage",
+        "value": 3.688,
+        "date": "2023-07-19T07:51:25.508306410Z"
+      },
+      "warnings": [
+        ""
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:58]:
```suggestion
        
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+      "recvTime": "2023-07-19T07:51:25.508306410Z"
+    },
+    "output": {
+      "data": {
+        "variable": "humidity",
+        "value": 37.21,
+        "date": "2023-07-19T07:51:25.508306410Z"
+      },
+      "warnings": [
+        ""
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:87]:
```suggestion
        
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+      "recvTime": "2023-07-19T07:51:25.508306410Z"
+    },
+    "output": {
+      "data": {
+        "variable": "pin_state",
+        "value": true,
+        "date": "2023-07-19T07:51:25.508306410Z"
+      },
+      "warnings": [
+        ""
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:115]:
```suggestion
        
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+      "fPort": 125,
+      "recvTime": "2023-07-19T07:51:25.508306410Z"
+    },
+    "output": {
+      "data": {
+        "variable": "counter",
+        "value": 0,
+        "date": "2023-07-19T07:51:25.508306410Z"
+      },
+      "warnings": [""]
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:145]:
```suggestion
      "warnings": []
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+      "fPort": 125,
+      "recvTime": "2023-07-19T07:51:25.508306410Z"
+    },
+    "output": {
+      "data": {
+        "variable": "configure reporting response status",
+        "value": 0,
+        "date": "2023-07-19T07:51:25.508306410Z"
+      },
+      "warnings": [""]
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:171]:
```suggestion
      "warnings": []
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+      "recvTime": "2023-07-19T07:51:25.508306410Z"
+    },
+    "output": {
+      "data": {
+        "variable": "read reporting configuration response status",
+        "value": 0,
+        "date": "2023-07-19T07:51:25.508306410Z"
+      },
+      "warnings": [
+        ""
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:208]:
```suggestion
        
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+      "recvTime": "2023-07-19T07:51:25.508306410Z"
+    },
+    "output": {
+      "data": {
+        "variable": "read reporting configuration response status",
+        "value": "no data",
+        "date": "2023-07-19T07:51:25.508306410Z"
+      },
+      "warnings": [
+        ""
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:241]:
```suggestion
        
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+          "value": 37,
+          "date": "2023-07-18T12:19:53.243Z"
+        },
+        {
+          "variable": "battery_voltage",
+          "value": 3.707,
+          "date": "2023-07-18T12:19:52.243Z"
+        }
+      ],
+      "warnings": [""]
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:314]:
```suggestion
      "warnings": []
```

**Review by mostafaibr — CHANGES_REQUESTED** [review]

**Comment by mostafaibr** [comment]
Please do not hesitate to get back to us if you have any doubt about the changes to do, we could organize a short session to fix all the issues together.

File: vendors/watteco/drivers/decode.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+                batch1: batch_parameters[0],
+                batch2: batch_parameters[1],
+                payload: payload,
+                date: date,
+            }
+            try {
+                let decoded = batch.normalisation_batch(batchInput)
+                return {
+                    data: decoded,
+                    warnings: [""],
```
Comment (mostafaibr) [inline vendors/watteco/drivers/decode.js:23]:
```suggestion
                    warnings: [],
```

File: vendors/watteco/drivers/decode.js [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        } else {
+            return {
+                data: decoded.data,
+                warnings: [decoded.warning],
+            };
+        }
+    } catch (error) {
+        return {
+            error: error.message,
+            warnings: [""],
```
Comment (mostafaibr) [inline vendors/watteco/drivers/decode.js:40]:
```suggestion
            warnings: [],
```

File: vendors/watteco/drivers/th_v4/examples.json [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+      "fPort": 125,
+      "recvTime": "2023-07-19T07:51:31.598957793Z"
+    },
+    "output": {
+      "data": {
+        "variable": "temperature",
+        "value": 30.91,
+        "date": "2023-07-19T07:51:31.598957793Z"
+      },
+      "warnings": [["alarm triggered"]]
```
Comment (mostafaibr) [inline vendors/watteco/drivers/th_v4/examples.json:269]:
```suggestion
      "warnings": ["alarm triggered"]
```

**Comment by mostafaibr** [comment]
Hello @Logia556 ,

I wanted to apply the suggested changes on your PR, but did not have the access.

I have created a new PR so you can check it: https://github.com/actility/device-catalog/pull/61

In that case, no need to fix anything from your side, if you agree on my PR, we could directly merge it and close this one.
A new version of device profiles and drivers would be released before the end of the week. so If you want to include your modifications, please take a look on [this PR](https://github.com/actility/device-catalog/pull/61) and get back to me as soon as possible. 

Thank you

**Pull request closed without merging**